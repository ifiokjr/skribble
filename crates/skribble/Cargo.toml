[package]
name = "skribble"
version = { workspace = true }
authors = ["Ifiok Jr. <ifiokotung@gmail.com>"]
categories = { workspace = true }
documentation = "https://docs.rs/skribble"
edition = { workspace = true }
homepage = { workspace = true }
include = ["src/**/*.rs", "Cargo.toml", "readme.md"]
keywords = ["build", "atomic", "css"]
license = { workspace = true }
readme = "readme.md"
repository = { workspace = true }
rust-version = { workspace = true }
description = "skribble is a fully typed atomic css compiler for rust, node, deno and the browser"

[dependencies]
doc-comment = { workspace = true }
skribble_core = { workspace = true }
