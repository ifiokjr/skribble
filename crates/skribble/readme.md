# skribble

> a fully typed atomic css compiler for rust, node, deno and the browser
