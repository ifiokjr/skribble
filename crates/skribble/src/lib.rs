#![deny(clippy::all)]
#![forbid(clippy::indexing_slicing)]

doc_comment::doctest!("../readme.md");

use std::path::Path;

pub use skribble_core as core;
pub use skribble_core::vfs;
pub use skribble_preset as preset;
pub use skribble_rust as rust;

pub use crate::core::Error;
use crate::core::PluginContainer;
pub use crate::core::Result;
use crate::core::SkribbleRunner;
use crate::core::StyleConfig;
use crate::core::VfsPath;
use crate::preset::PresetPlugin;
use crate::rust::RustPlugin;

/// Create a default `StyleConfig` with the `PresetPlugin` and `RustPlugin`
/// builtin.
///
/// The plugins can be overriden since the `StyleConfig` can be mutated with
/// methods like `config.add_plugins()` and `config.remove_plugin()`.
pub fn create_config() -> StyleConfig {
  let default_preset = PresetPlugin::builder().build();
  let rust_plugin = RustPlugin::builder().build();

  let config: StyleConfig = StyleConfig::builder()
    .plugins(vec![
      PluginContainer::from(default_preset),
      PluginContainer::from(rust_plugin),
    ])
    .build();

  config
}

/// Generate the files that created by plugins and also the css files.
pub fn run_with_config(config: StyleConfig) -> Result<SkribbleRunner> {
  let mut runner = SkribbleRunner::try_new(config)?;
  let _ = runner.initialize()?;
  let _generate_files = runner.generate()?; // Write the files to their destinations.
  let _css_result = runner.scan()?; // Write the css file to the destination.

  Ok(runner)
}

/// Generate the files that created by plugins and also the css files.
///
/// This is likely to change a lot in the future.
pub fn run(
  config: StyleConfig,
  cwd: impl AsRef<Path>,
  vfs: Option<VfsPath>,
) -> Result<SkribbleRunner> {
  let mut runner = SkribbleRunner::new(config, cwd, vfs);
  let _ = runner.initialize()?;
  let mut generated_files = runner.generate()?;
  let css_result = runner.scan()?;

  // Write the files to their destinations.
  runner.write_files(&mut generated_files)?;
  // Write the css file to the destination.
  runner.write_css(&css_result)?;

  Ok(runner)
}

#[cfg(test)]
mod __tests;
