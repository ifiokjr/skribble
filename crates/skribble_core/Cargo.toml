[package]
name = "skribble_core"
version = { workspace = true }
authors = { workspace = true }
categories = { workspace = true }
documentation = "https://docs.rs/skribble_core"
edition = { workspace = true }
homepage = { workspace = true }
include = ["src/**/*.rs", "Cargo.toml", "readme.md"]
keywords = ["build", "atomic", "css"]
license = { workspace = true }
readme = "readme.md"
repository = { workspace = true }
rust-version = { workspace = true }
description = "Core features for the skribble atomic css library"

[dependencies]
derivative = { workspace = true }
derive_more = { workspace = true }
doc-comment = { workspace = true }
globset = { workspace = true }
heck = { workspace = true }
indent_write = { workspace = true }
indexmap = { features = ["serde"], workspace = true }
lazy_static = { workspace = true }
lightningcss = { workspace = true }
readonly = { workspace = true }
regex = { workspace = true }
serde = { workspace = true }
serde_json = { workspace = true }
skribble_color = { workspace = true, features = ["serde"] }
thiserror = { workspace = true }
typed-builder = { workspace = true }
vfs = { workspace = true }

[dev-dependencies]
insta = { features = ["json"], workspace = true }
