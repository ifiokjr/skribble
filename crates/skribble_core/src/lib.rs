#![deny(clippy::all)]

pub use error::*;

mod error;
