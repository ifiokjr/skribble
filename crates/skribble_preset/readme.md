# skribble_preset

<br />

[![Crate][crate-image]][crate-link] [![Docs][docs-image]][docs-link] [![Status][ci-status-image]][ci-status-link] [![Unlicense][unlicense-image]][unlicense-link]

## Installation

```toml
[dependencies]
skribble_preset = "0.0.0"
```

### Usage

```rust
use skribble_preset::*;
```

[crate-image]: https://img.shields.io/crates/v/skribble_preset.svg
[crate-link]: https://crates.io/crates/skribble_preset
[docs-image]: https://docs.rs/skribble_preset/badge.svg
[docs-link]: https://docs.rs/skribble_preset
[ci-status-image]: https://github.com/ifiokjr/skribble/workflows/ci/badge.svg
[ci-status-link]: https://github.com/ifiokjr/skribble/actions?query=workflow:ci
[unlicense-image]: https://img.shields.io/badge/license-Unlicence-blue.svg
[unlicense-link]: https://opensource.org/license/unlicense
