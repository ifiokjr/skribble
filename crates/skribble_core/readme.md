# skribble_core

> core library for skribble

<br />

[![Crate][crate-image]][crate-link] [![Docs][docs-image]][docs-link] [![Status][ci-status-image]][ci-status-link] [![Unlicense][unlicense-image]][unlicense-link]

## Installation

```toml
[dependencies]
skribble_core = "0.0.0"
```

### Usage

```rust
use skribble_core::*;
```

[crate-image]: https://img.shields.io/crates/v/skribble_core.svg
[crate-link]: https://crates.io/crates/skribble_core
[docs-image]: https://docs.rs/skribble_core/badge.svg
[docs-link]: https://docs.rs/skribble_core
[ci-status-image]: https://github.com/ifiokjr/skribble/workflows/ci/badge.svg
[ci-status-link]: https://github.com/ifiokjr/skribble/actions?query=workflow:ci
[unlicense-image]: https://img.shields.io/badge/license-Unlicence-blue.svg
[unlicense-link]: https://opensource.org/license/unlicense
