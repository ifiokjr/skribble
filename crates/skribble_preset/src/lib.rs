#![deny(clippy::all)]
#![forbid(clippy::indexing_slicing)]

doc_comment::doctest!("../readme.md");

use data::*;
use enums::*;
use serde::Deserialize;
use serde::Serialize;
use skribble_core::*;
use typed_builder::TypedBuilder;

pub(crate) mod data;
mod enums;

#[derive(Debug, Clone, Default, Deserialize, TypedBuilder, Serialize)]
pub struct PresetDefault {
  /// Choose the palette colors from either Tailwind or OpenColors.
  #[builder(default, setter(into))]
  pub palette: PaletteType,
  /// Choose how `light` and `dark` mode are handled. via `prefers-color-scheme`
  /// or `class` based.
  #[builder(default, setter(into))]
  pub dark_mode: DarkMode,
  /// Choose to ignore the color variables so that you can define your own
  /// custom theme.
  #[builder(default, setter(into))]
  pub ignore_colors: bool,
}

impl Plugin for PresetDefault {
  fn get_data(&self) -> PluginData {
    PluginData::builder()
      .id("skribble_preset")
      .name("Default Preset")
      .description(
        "This plugin provides a default preset for Skribble which is similar to `tailwind`, \
         `windi` and `unocss`.",
      )
      .build()
  }

  fn mutate_config(&mut self, config: &mut PluginConfig, _options: &Options) -> AnyEmptyResult {
    self.update_palette(&mut config.palette);
    self.update_media_queries(&mut config.media_queries);
    self.update_modifiers(&mut config.modifiers);
    self.update_variables(&mut config.variables);
    self.update_keyframes(&mut config.keyframes);
    self.update_atoms(&mut config.atoms);
    self.update_named_classes(&mut config.classes);
    self.update_value_sets(&mut config.value_sets);

    Ok(())
  }
}

impl PresetDefault {
  fn update_media_queries(&self, media_queries: &mut MediaQueries) {
    media_queries.extend(MEDIA_QUERIES.clone());

    if self.dark_mode == DarkMode::Media {
      media_queries.extend(DARK_MEDIA_QUERIES.clone());
    }
  }

  fn update_palette(&self, palette: &mut Palette) {
    palette.extend(self.palette.palette());
  }

  fn update_modifiers(&self, modifiers: &mut Modifiers) {
    modifiers.extend(MODIFIERS.clone());

    if self.dark_mode == DarkMode::Class {
      modifiers.extend_group(DARK_PARENT_MODIFIERS.clone());
    }
  }

  fn update_variables(&self, css_variables: &mut CssVariables) {
    if !self.ignore_colors {
      css_variables.extend(COLOR_CSS_VARIABLES.clone());
    }

    css_variables.extend(CSS_VARIABLES.clone());
  }

  fn update_keyframes(&self, keyframes: &mut Keyframes) {
    keyframes.extend(KEYFRAMES.clone());
  }

  fn update_atoms(&self, atoms: &mut Atoms) {
    atoms.extend(ATOMS.clone());
  }

  fn update_named_classes(&self, named_classes: &mut NamedClasses) {
    named_classes.extend(NAMED_CLASSES.clone());
  }

  fn update_value_sets(&self, value_sets: &mut ValueSets) {
    value_sets.extend(ATOM_VALUE_SETS.clone());
    value_sets.extend(ANIMATION_VALUE_SETS.clone());
  }
}

#[cfg(test)]
mod __tests;
