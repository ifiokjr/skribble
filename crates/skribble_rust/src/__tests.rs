use skribble_core::vfs::MemoryFS;
use skribble_core::*;
use skribble_preset::PresetPlugin;

use super::*;

#[test]
fn can_generate_skribble_rust_code() -> AnyEmptyResult {
  let default_preset = PresetPlugin::builder().build();
  let rust_plugin = RustPlugin::builder().build();

  let config: StyleConfig = StyleConfig::builder()
    .plugins(vec![
      PluginContainer::from(default_preset),
      PluginContainer::from(rust_plugin),
    ])
    .build();

  let vfs: VfsPath = create_memory_fs()?;
  let mut runner = SkribbleRunner::new(config, "/", Some(vfs));
  let _ = runner.initialize()?;
  let result = runner.generate()?;
  let GeneratedFile { content, .. } = result.first().ok_or(Error::Unknown)?;
  insta::assert_display_snapshot!(content);

  Ok(())
}

#[test]
fn can_scan_and_generate_css() -> AnyEmptyResult {
  let default_preset = PresetPlugin::builder().build();
  let rust_plugin = RustPlugin::builder().build();

  let config: StyleConfig = StyleConfig::builder()
    .plugins(vec![
      PluginContainer::from(default_preset),
      PluginContainer::from(rust_plugin),
    ])
    .build();

  let vfs: VfsPath = create_memory_fs()?;
  let mut runner = SkribbleRunner::new(config, "/", Some(vfs));
  let _ = runner.initialize()?;
  let scanned = runner.scan()?;
  insta::assert_display_snapshot!(scanned.code);

  Ok(())
}

fn create_memory_fs() -> AnyResult<VfsPath> {
  let vfs: VfsPath = MemoryFS::new().into();

  for file in FILES {
    let path = vfs.join(file.0)?;
    path.create_dir_all()?;
    let mut writer = path.create_file()?;
    write!(writer, "{}", file.1)?;
  }

  Ok(vfs)
}

const FILES: &[(&str, &str)] = &[
  (
    "src/lib.rs",
    r#"
use leptos::*;
use crate::skribble::*;
#[component]
fn App(cx: Scope) -> impl IntoView {
  let s = sk().md().p().px();
  let a = sk().bg().secondary();
  let classes = vec![sk().bg().secondary(), sk().p().px()].join(" ");

  view! {
    cx,
    <div class={classes}>
      <h1 class={sk().text().color().primary()}>Hello World</h1>
    </div>
  }
}

mod other;
"#,
  ),
  (
    "src/other.rs",
    r#"
use crate::skribble::*;
pub fn other() -> String {
  sk().bg().primary()
}
  "#,
  ),
];
