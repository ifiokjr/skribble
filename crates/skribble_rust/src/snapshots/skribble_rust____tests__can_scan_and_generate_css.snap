---
source: crates/skribble_rust/src/__tests.rs
expression: scanned.code
---
@layer default;

@property --sk-opacity-s {
  syntax: "<number>";
  inherits: true;
  initial-value: 1;
}

@property --sk-s {
  syntax: "<color>";
  inherits: true;
  initial-value: #f000b8;
}

@property --sk-opacity-p {
  syntax: "<number>";
  inherits: true;
  initial-value: 1;
}

@property --sk-p {
  syntax: "<color>";
  inherits: true;
  initial-value: #570df8;
}

@layer default {
  :root {
    --sk-opacity-s: 1;
    --sk-s: hsl(314 100% 47.0588% / var(--sk-opacity-s));
    --sk-opacity-p: 1;
    --sk-p: hsl(258.894 94.3775% 51.1765% / var(--sk-opacity-p));
  }

  .bg\:\$secondary {
    --sk-opacity-s: 1;
    background-color: var(--sk-s);
  }

  .bg\:\$primary {
    --sk-opacity-p: 1;
    background-color: var(--sk-p);
  }

  @media (min-width: 768px) {
    .md\:p\:\$px {
      padding: 1px;
    }
  }
}

